File: services/typescript/src/service/project-service.ts
import { InvalidInputError } from "../repository/errors.js";
import {
  ProjectRepo,
  type CreateProjectParams,
  type ListProjectsParams,
  type Project,
  type ProjectList,
  type UpdateProjectParams,
} from "../repository/project-repo.js";
import type { Logger } from "pino";

export class ProjectService {
  constructor(
    private repo: ProjectRepo,
    private logger: Logger,
  ) {}

  async create(params: CreateProjectParams): Promise<Project> {
    if (!params.name) {
      throw new InvalidInputError("invalid input: name is required");
    }
    if (!params.workspaceId) {
      throw new InvalidInputError("invalid input: workspace_id is required");
    }
    this.logger.debug({ name: params.name, workspaceId: params.workspaceId }, "creating project");
    return this.repo.create(params);
  }

  async getById(id: string): Promise<Project> {
    return this.repo.getById(id);
  }

  async list(params: ListProjectsParams): Promise<ProjectList> {
    return this.repo.list(params);
  }

  async update(params: UpdateProjectParams): Promise<Project> {
    if (!params.name) {
      throw new InvalidInputError("invalid input: name is required");
    }
    const validStatuses: Record<string, boolean> = { active: true, archived: true };
    if (params.status && !validStatuses[params.status]) {
      throw new InvalidInputError(`invalid input: invalid status: ${params.status}`);
    }
    return this.repo.update(params);
  }

  async delete(id: string): Promise<void> {
    return this.repo.delete(id);
  }
}

File: services/kotlin/src/main/kotlin/com/corestack/service/ProjectService.kt
package com.corestack.service

import com.corestack.repository.*
import io.github.oshai.kotlinlogging.KotlinLogging
import java.util.UUID

private val logger = KotlinLogging.logger {}

class ProjectService(private val repo: ProjectRepo) {

    suspend fun create(params: CreateProjectParams): Project {
        if (params.name.isBlank()) throw InvalidInputError("invalid input: name is required")
        if (params.workspaceId == UUID(0, 0)) throw InvalidInputError("invalid input: workspace_id is required")
        logger.debug { "creating project name=${params.name} workspace_id=${params.workspaceId}" }
        return repo.create(params)
    }

    suspend fun getById(id: UUID): Project = repo.getById(id)

    suspend fun list(params: ListProjectsParams): ProjectList = repo.list(params)

    suspend fun update(params: UpdateProjectParams): Project {
        if (params.name.isBlank()) throw InvalidInputError("invalid input: name is required")
        val validStatuses = setOf("active", "archived")
        if (params.status.isNotEmpty() && params.status !in validStatuses) {
            throw InvalidInputError("invalid input: invalid status: ${params.status}")
        }
        return repo.update(params)
    }

    suspend fun delete(id: UUID) = repo.delete(id)
}
